src/PageStreamingDescriptor.php

```diff
@@ -38,61 +38,89 @@
  */
 class PageStreamingDescriptor
 {
-    private $requestPageTokenField;
-    private $requestPageSizeField;
-    private $responsePageTokenField;
-    private $resourceField;
-
-    /**
-     * @param array $descriptor {
-     *     Required.
-     *
-     *     @type string $requestPageTokenField the page token field in the request object.
-     *     @type string $responsePageTokenField the page token field in the response object.
-     *     @type string $resourceField the resource field in the response object.
-     * }
-     */
+    private $descriptor;
+
     public function __construct($descriptor)
     {
         self::validate($descriptor);
-        $this->requestPageTokenField = $descriptor['requestPageTokenField'];
-        if (isset($descriptor['requestPageSizeField'])) {
-            $this->requestPageSizeField = $descriptor['requestPageSizeField'];
-        } else {
-            $this->requestPageSizeField = null;
+        $this->descriptor = $descriptor;
+    }
+
+    public static function createFromFields($fields)
+    {
+        $requestPageToken = $fields['requestPageTokenField'];
+        $responsePageToken = $fields['responsePageTokenField'];
+        $resources = $fields['resourceField'];
+
+        $descriptor = [
+            'requestPageTokenGetMethod' => PageStreamingDescriptor::getMethod($requestPageToken),
+            'requestPageTokenSetMethod' => PageStreamingDescriptor::setMethod($requestPageToken),
+            'responsePageTokenGetMethod' => PageStreamingDescriptor::getMethod($responsePageToken),
+            'resourcesGetMethod' => PageStreamingDescriptor::getMethod($resources),
+        ];
+
+        if (isset($fields['requestPageSizeField'])) {
+            $requestPageSize = $fields['requestPageSizeField'];
+            $descriptor['requestPageSizeGetMethod'] = PageStreamingDescriptor::getMethod($requestPageSize);
+            $descriptor['requestPageSizeSetMethod'] = PageStreamingDescriptor::setMethod($requestPageSize);
         }
-        $this->responsePageTokenField = $descriptor['responsePageTokenField'];
-        $this->resourceField = $descriptor['resourceField'];
+
+        return new PageStreamingDescriptor($descriptor);
+    }
+
+    private static function getMethod($field)
+    {
+        return 'get' . ucfirst($field);
     }
 
-    public function getRequestPageTokenField()
+    private static function setMethod($field)
     {
-        return $this->requestPageTokenField;
+        return 'set' . ucfirst($field);
     }
 
-    public function getRequestPageSizeField()
+    public function getRequestPageTokenGetMethod()
     {
-        return $this->requestPageSizeField;
+        return $this->descriptor['requestPageTokenGetMethod'];
```

Comment:
What's the advantage of having a key/value array instead of storing this data as member variables?